Add new conditionalHnsRestarter.p1 script

jayanthAP opened the pull request:
Script to be used to restart HNS based when certain condition is met, to help mitigate issues in customer AKS nodes. Script currently supports checking for missing rules on pod-endpoints.

Example command:
conditionalHnsRestarter.ps1 -PodPortRulesToCheck @{"ruleRegex" = "LB_DSR_[0-9A-Z]+_[\d\.]+_[\d\.]+_801_801_17"; "layerName" = "LB_DSR"; "groupName" = "LB_DSR_IPv4_OUT"}

jayanthAP committed:
Create conditionalHnsRestarter.p1

Script to be used to restart HNS based when certain condition is met, to help mitigate issues in customer AKS nodes. Script currently supports checking for missing rules on pod-endpoints.

Example command:
conditionalHnsRestarter.ps1 -PodPortRulesToCheck @{"ruleRegex" = "LB_DSR_[0-9A-Z]+_[\d\.]+_[\d\.]+_801_801_17"; "layerName" = "LB_DSR"; "groupName" = "LB_DSR_IPv4_OUT"}

jayanthAP committed:
Rename conditionalHnsRestarter.p1 to conditionalHnsRestarter.ps1

jayanthAP committed:
Update and rename conditionalHnsRestarter.ps1 to conditionalMitigator.ps1

This is a generic mitigator script. Currently it checks pod port rules and restarts HNS when rules are not found. It can be extended for other mitigation steps as well, since the mitigation logic can be similar.

jayanthAP committed:
Update conditionalMitigator.ps1 2

jayanthAP committed:
Update conditionalMitigator.ps1 3

princepereira commented on scripts/conditionalMitigator/conditionalMitigator.ps1:
```diff
@@ -0,0 +1,409 @@
+        #$msg = "Checking for rules on port name:{0} id:{1}" -f $g_currentVfpPortMap[$portId].name,$g_currentVfpPortMap[$portId].id
+        #LogWithTimeStamp -msgStr $msg
+
+        $rulesPresent = RulesPresentOnVfpPort -portId $portId -rulesToCheck $g_podRuleCheckList
```
RulesPresentOnVfpPort: function name is misleading. Give something like : CheckAllRulesPresentOnVfpPort or IsAllRulesPresentOnVfpPort

princepereira commented on scripts/conditionalMitigator/conditionalMitigator.ps1:
```diff
@@ -0,0 +1,409 @@
+        #$msg = "Layer {0} found on port {1}: {2}." -f $ruleToCheck.layerName, $portId, $layers[$layerIndex]
+        #LogWithTimeStamp -msgStr $msg
+
+        $rulePresentInLayer = RulePresentInVfpPortLayer -layer $layers[$layerIndex] -ruleToCheck $ruleToCheck
```
IsRulePresentInVfpPortLayer

princepereira commented on scripts/conditionalMitigator/conditionalMitigator.ps1:
```diff
@@ -0,0 +1,409 @@
+    [RuleCheckInfo] $ruleToCheck
+)
+{
+    # first find layer
```
Copy paste comment: should be "Find group"

princepereira commented on scripts/conditionalMitigator/conditionalMitigator.ps1:
```diff
@@ -0,0 +1,409 @@
+        return $false
+    }
+
+    #$msg = "group {0} found in layer {1}." -f $ruleToCheck.groupName, $ruleToCheck.layerName
```
remove unwanted comments

princepereira commented on scripts/conditionalMitigator/conditionalMitigator.ps1:
```diff
@@ -0,0 +1,409 @@
+        }
+        elseif ($g_mitigationActionCount -ge $MaxMitigationCount)
+        {
+            $msg = "Not taking mitigation-action since MaxMitigationCount has been crossed. Shouldn't take action anymore."
```
In this cases, you can avoid further processing. Just keep looping here

princepereira commented on scripts/conditionalMitigator/conditionalMitigator.ps1:
```diff
@@ -0,0 +1,409 @@
+)
+{
+    # create log path if not yet created.
+    mkdir -Force $LogsPath
```
You can move mkdir under if condition

princepereira commented on scripts/conditionalMitigator/conditionalMitigator.ps1:
```diff
@@ -0,0 +1,409 @@
+param (
+    [Parameter(Mandatory=$true)]
```
If makes sense, you can also add this under a flag


**At the beginning of script:**

Stop capture which is already running: pktmon stop

Start Capture: pktmon start --trace -p Microsoft-Windows-Host-Network-Service -l 6 -m circular -s 1000

Repro:

**Once issue is identified:**

Stop capture: pktmon stop

Convert capture (inside the same node): pktmon etl2txt pktmon.etl

jayanthAP replied:
Do you mean using the script to start and stop pktmon?

princepereira replied:
Yes, in circular mode with 1 GB as log size.

jayanthAP replied:
Currently, not planning on collecting trace logs. Once we have better trace logging support in HNS at appropriate level, we can plan on collecting trace logs. Another thing is, enabling trace logs at daemonset level could be an issue. Need to understand the memory/disk cost of doing so.

princepereira commented on scripts/conditionalMitigator/conditionalMitigator.ps1:
```diff
@@ -0,0 +1,409 @@
+        ####
+        # Conditions for not mitigating.
+        if ($scriptAge.TotalMinutes -lt $SleepIntervalMins)
+        {
```
You can avoid this condition and make the script sleep immediately after the mitigation action is taken

princepereira commented on scripts/conditionalMitigator/conditionalMitigator.ps1:
```diff
@@ -0,0 +1,409 @@
+    #LogWithTimeStamp -msgStr $msg
+
+    ## Note new endpoint IDs.
+    $priorSize = $g_endpointInfoMap.count
```
Add a comment saying: Adding new endpoints to $g_endpointInfoMap

princepereira commented on scripts/conditionalMitigator/conditionalMitigator.ps1:
```diff
@@ -0,0 +1,409 @@
+
+
+    ## Delete stale endpoint IDs, so that g_endpointInfoMap's size does not keep increasing forever.
+    $stalePortIdList = @()
```
Add a comment saying: Removing deleted endpoints from $g_endpointInfoMap

jayanthAP committed:
Update conditionalMitigator.ps1 4

jayanthAP committed:
Update conditionalMitigator.ps1 5

jayanthAP committed:
Update conditionalMitigator.ps1 6

jayanthAP committed:
Update conditionalMitigator.ps1 6

jayanthAP committed:
Update conditionalMitigator.ps1 8

jayanthAP committed:
Update conditionalMitigator.ps1 9

princepereira commented on scripts/conditionalMitigator/conditionalMitigator.ps1:
```diff
@@ -0,0 +1,391 @@
+{
+    if ($CollectWindowsLogs -eq $true) {
+        # create log path if not yet created.
+        mkdir -Force $LogsPath
```
If $CollectWindowsLogs is $true and issue happens very frequently, there are chances multiple log files will be created. This will consume lot of customer's space.

jayanthAP replied:
Yes, we will need to add a rotation logic for the same. Will add that.

jayanthAP committed:
Update conditionalMitigator.ps1 10

jayanthAP committed:
Update conditionalMitigator.ps1 11

jayanthAP committed:
Update conditionalMitigator.ps1 12

jayanthAP committed:
Create crashEventChecker.yaml

hnsCrashEventChecker.yaml is a daemonset that looks for HNS crash win-event logs. Once applied on a kubernetes cluster, user can execute below command to know the timestamps when HNS crashed occurred on respective windows nodes:

kubectl logs -l name=hns-crash-event-checker --all-containers=true

jayanthAP committed:
Rename hnsCrashEventChecker.yaml to crashEventChecker.yaml

Making the daemonset generic such that it can be used for any service (not just HNS). The $ServiceName variable needs to be edited for this.

jayanthAP committed:
Correct crashEventChecker.yaml

Last edit missed the daemonset code.

jayanthAP committed:
Update crashEventChecker.yaml

jayanthAP committed:
Create CrashDumpEnabler

This daemonset enables collection of crash-dumps for user-space processes on AKS windows nodes.

jayanthAP committed:
Rename scripts/CrashDumpEnabler to scripts/crashDumpEnabler/CrashDumpEnabler.yaml

daschott commented:
Is this ready for review / merging?
